test-suite/test-functionality.sh

```diff
@@ -0,0 +1,264 @@
+    then
+        echo "Will reuse squid-overlord service running at $url"
+        return 0;
+    fi
+
+    # Do not be tempted to simply run `sudo ... overlord.pl`: User nobody will
+    # lack read permissions, and sudo will ask for a password.
+    sudo -n --background -u nobody perl < $SQUID_OVERLORD_DIR/overlord.pl > $log 2>&1 || return
+    echo "Started squid-overlord service at $url"
+}
+
+setup_test_tools() {
+    echo "::group::Setup test tools"
+
+    clone_repo https://github.com/measurement-factory/daft $DAFT_DIR || return
+    clone_repo https://github.com/measurement-factory/squid-dafts $SQUID_DAFTS_DIR || return
+    clone_repo https://github.com/measurement-factory/squid-overlord $SQUID_OVERLORD_DIR || return
+
+    if ! test -e $SQUID_DAFTS_DIR/src
+    then
+        run ln -s `realpath $DAFT_DIR/src` $SQUID_DAFTS_DIR/src || return
+    fi
+    if ! test -e $SQUID_DAFTS_DIR/node_modules
+    then
+        run ln -s `realpath $DAFT_DIR/node_modules` $SQUID_DAFTS_DIR/node_modules || return
+    fi
+
+    start_overlord || return
+
+    # TODO: Find a good way to end group on (currently fatal) errors as well.
+    echo "::endgroup::"
+}
+
+# executes a single test after the caller confirms that the test is applicable
+run_confirmed_test() {
+    local testId="$1"
+
+    local testRunner="$DAFT_DIR/src/cli/daft.js"
+    if ! test -e $testRunner
+    then
+        echo_error "Missing Daft test execution script"
+        echo "Expected to find it at $testRunner"
+        exit 1;
+    fi
+
+    local testsDir="$SQUID_DAFTS_DIR/tests"
+    if ! test -d $testsDir
+    then
+        echo_error "Missing collection of Squid-specific Daft tests"
+        echo "Expected to find them in $testsDir/"
+        exit 1;
+    fi
+
+    local testScript=$testsDir/$testId.js
+    if ! test -e $testScript
+    then
+        echo_error "Unknown test requested: $testId"
+        echo "Expected to find it at $testScript"
+        return 1;
+    fi
+
+    local log="$TMPDIR/$testId.log"
+
+    echo "Running test: $testId"
+    local result=undefined
+    if $testRunner run $testScript > $log 2>&1
+    then
+        echo "Test $testId: OK"
+        return 0;
+    else
+        result=$?
+    fi
+
+    # TODO: Report skipped tests and ignored failures more prominently. See
+    # test-sources.sh for CHECK_OUTCOME_PHRASE tricks (but avoid duplication).
+    echo
+    echo_error "Test $testId: Failed with exit code $result"
+    echo "::group::Test log tail:"
+    tail -n 100 $log
+    echo "::endgroup::"
+
+    # TODO: Link to the artifact
+    echo "See the test log (tailed above) for failure details: $log"
+    return $result
+}
+
+check_pconn() {
+    run_confirmed_test pconn
+}
+
+check_busy_restart() {
+    if ! run_confirmed_test busy-restart
+    then
+        # XXX: Make the test stable instead!
+        echo_warning "Ignoring unstable test failure: busy-restart"
```

Comment:
This PR was motivated, in part, by Amos request to make Semaphore CI tests more stable. AFAICT, most of Sempaphore CI failures could be attributed to two things:

* Problems/complications with the overall Semaphore CI setup. We are _not_ inheriting those problems here.
* An unstable proxy-collapsed-forwarding test. That test is stable enough in Factory lab, but not in Sempahore CI environment. It appears to be stable enough in Github Actions environment (which is much closer to what we use in the lab), which is the outcome I was hoping for.

This busy-restart test was stable in the Semaphore CI environment. It is stable in our lab as well. I am surprised it usually fails in Github Actions environment. Factory will work on making it stable in Github Actions environment. This test is not as complex as the proxy-collapsed-forwarding, and I am more optimistic that we can find a way to make its caching less susceptible to various natural race conditions/problems.

For now, I think it is best to keep it running but ignore its failures. The PR does that. If others insist, I will remove it instead.